### PR TITLE
Pointers In C#

### LINKED ISSUE #1987: Pointers in c#
### Proposed title of the article
Pointers in c#

### Introduction paragraph (2-3 paragraphs):

C# upholds pointers to a restricted degree. A C# pointer is only a variable that holds the memory address of another kind. In any case, the C# pointer must be announced to hold the memory address of significant worth sorts and clusters.

 Unlike reference types, pointer types are not followed by the default trash assortment system. For a similar explanation, pointers are not permitted to highlight a reference type or even to a construction type that contains a reference type. We can say that pointers can highlight just unmanaged types which incorporates all fundamental information types, enum types, other pointer types, and structs that contain just unmanaged types. 

In this article, we will understand what pointers are, how to declare a pointer in a C# program, and using the pointers with different operations such as arrays, methods, structure, and conversions.

### Key takeaways: 
in this article, the reader will be able to understand
- what are pointers.
- safe and unsafe codes.
- how to declare a pointer type.
- how to pin an object in C#.
- how to Perform different operations in C# using pointers.

### References:
N/A

### PR DESCRIPTION
**Pull request by carteblanchemunene**
### Article checklist

**ATTENTION: In your PR - add the ISSUE (topic form) # that this PR corresponds to. You can do this by also adding a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.**
      
      
**NOTE: (Please ensure that you have only one open issue + linked pull request at a time. This will ensure that we complete the article in a timely manner from inception to publishing.)**

**If you have not already please go over our [Resources Page](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md) for more tips and tricks on how to improve our overall technical writing.**

#### Preliminary Checks - Formatting and Structure 
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**
- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
- [x] In your PR - add a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) -  [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker? We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [ ] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**

### Contribution guidelines
For first-time contributors and for more details, **see [Contributing Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
this fixes #1987

### CODE REVIEW

**Comment by hectorkambow** [comment]
Thank you v much for your patience as we worked through our Repo re-structuring process. 

The purpose behind this restructure - was to better allow us to handle more volume and allow students to publish more articles and get more involved with the community (with open-source contribution and suggestions). 

With most of the changes coming to completion - we will be sure to get your article published asap. 

Thank you again - and we are looking forward to more amazing articles.

File: content/articles/pointers-in-c#/index.md.md [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+        public static void Main()
+        {
+            Demo d = new Demo();
+            d.Method();
+        }
+    }
+} 
+```
+
+![Output](/engineering-education/active-learning/unsafe%20codes.PNG)
```
Comment (ahmadmardeni1) [inline content/articles/pointers-in-c#/index.md.md:64]: The folder name is wrong. And no need for the image, please provide the output as a text.

File: content/articles/pointers-in-c#/index.md.md [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+        public static void Main()
+        {
+            Demo d = new Demo();
+            d.Method();
+        }
+    }
+}
+```
+
+![Output](/engineering-education/active-learning/pointers and methods .PNG)
```
Comment (ahmadmardeni1) [inline content/articles/pointers-in-c#/index.md.md:141]: Folder name is wrong, and the image name must not include spaces.

**Review by ahmadmardeni1 — CHANGES_REQUESTED** [review]
Hello 👋
As a start, welcome to the Section community, and congrats on your first tutorial!✨

At a quick glance, this seems to be an interesting article. A few changes and we can continue the primary revision then start the technical review:
1. Images folder name is wrong, fix it. (Some images are not in the right structure, take a look [here](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/UPLOAD_INSTRUCTIONS.md).
2. If the output is a text or a code then no need for a screenshot, just add it as a code block.
3. Add a table of contents.
4. Add a further reading section.

Ping me when you are done, and please, let me know (here or in Slack) if you need any help.

**Comment by hectorkambow** [comment]
plagiarism check done

**Comment by carteblanchemunene** [comment]
@ahmadmardeni1 changes done.

**Comment by ahmadmardeni1** [comment]
Please delete the unused pictures.
And link the "table of contents " contents to their headers. [Click here](https://stackoverflow.com/questions/11948245/markdown-to-create-pages-and-table-of-contents) if you don't know-how.

**Comment by carteblanchemunene** [comment]
@ahmadmardeni1 Changes done. Please check my last commit.

File: content/articles/pointers-in-c#/index.md [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+} 
+```
+
+And the output is:
+
+```
+60
+30
+1605887284
+1605887280
```
Comment (ahmadmardeni1) [inline content/articles/pointers-in-c#/index.md:83]: The reader should know that the last two results may be different in his machine based on the values addresses.

File: content/articles/pointers-in-c#/index.md [hunk 4]
```diff
@@ -0,0 +1,277 @@
...
+                Console.WriteLine("Address of array[{0}]={1}", k, (int)(ptr + k));
+            }
+            Console.ReadKey();
+        }
+
+    }
+
+}
+```
+![output](/engineering-education/pointers_in_c#/pinning_an_object.png)
```
Comment (ahmadmardeni1) [inline content/articles/pointers-in-c#/index.md:122]: No need for the picture. Please add the output as a code block.

File: content/articles/pointers-in-c#/index.md [hunk 5]
```diff
@@ -0,0 +1,277 @@
...
+                }
+            Console.ReadKey();
+        }
+
+    }
+
+}
+```
+
+![output](/engineering-education/pointers_in_c#/pointers_and_arrays.png)
```
Comment (ahmadmardeni1) [inline content/articles/pointers-in-c#/index.md:216]: No need for the picture. Please add the output as a code block.

**Review by ahmadmardeni1 — CHANGES_REQUESTED** [review]
This is a great tutorial, good job. 

Please pull my changes before making the changes I requested. And take a look at them to prevent those mistakes in future articles.

Ping me when you are done so we can move on to the final review.

**Comment by carteblanchemunene** [comment]
ahmadmardeni1
changes done.

**Comment by carteblanchemunene** [comment]
@ahmadmardeni1
changes done. Please check my last commit.

File: content/articles/pointers-in-c#/index.md [hunk 6]
```diff
@@ -0,0 +1,330 @@
...
+3. Pointer type to sbyte, byte, uint, int, long, ulong, short, ushort types.
+
+An example of pointer conversion is illustrated below:
+
+```c#
+char k = 'U';  
+char *pk = &k;  
+void *px = pk; 
+int *pj = (int *) px; 
+```
```
Comment (ahmadmardeni1) [inline content/articles/pointers-in-c#/index.md:221]: Exaplain this example. What conversions are happening here? (From what to what?)

**Review by ahmadmardeni1 — COMMENTED** [review]
This will be ready for a final review after you do the requested changes. @carteblanchemunene (Make sure you pulled my changes before making any commit)

Please clean your code. For example:
- Instead of declaring every variable in a separated line, declare them in one line.
- Instead of using multiple `Console.WriteLine();`. Use just one in every code block. Etc ...

**Pull request merged**